core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -2345,28 +2345,23 @@ private[spark] class RedirectThread(
  */
 private[spark] class CircularBuffer(sizeInBytes: Int = 10240) extends java.io.OutputStream {
   var pos: Int = 0
-  var buffer = new Array[Int](sizeInBytes)
+  var isBufferFull = false
+  var buffer = new Array[Byte](sizeInBytes)
 
-  def write(i: Int): Unit = {
-    buffer(pos) = i
+  def write(input: Int): Unit = {
+    buffer(pos) = input.toByte
     pos = (pos + 1) % buffer.length
+    isBufferFull = isBufferFull || (pos == 0)
   }
 
   override def toString: String = {
-    val (end, start) = buffer.splitAt(pos)
-    val input = new java.io.InputStream {
-      val iterator = (start ++ end).iterator
+    val end = new String(buffer, 0, pos, StandardCharsets.UTF_8)
```

Comment:
There could be a multibyte char encoding broken across end and start. I think we have to unfortunately build the whole byte array to convert and then read it as a string.